File: python/gherkin/pickles/compiler.py
import re

from ..count_symbols import count_symbols


def compile(gherkin_document, path):
    pickles = []
    if 'feature' not in gherkin_document:
        return pickles

    feature = gherkin_document['feature']
    feature_tags = feature['tags']
    background_steps = []
    for scenario_definition in feature['children']:
        args = (feature_tags, background_steps, scenario_definition, path, pickles)
        if scenario_definition['type'] is 'Background':
            background_steps = _pickle_steps(scenario_definition, path)
        elif scenario_definition['type'] is 'Scenario':
            _compile_scenario(*args)
        else:
            _compile_scenario_outline(*args)
    return pickles


def _compile_scenario(feature_tags, background_steps, scenario,
                      path, pickles):
    if len(scenario['steps']) == 0:
      return

    steps = list(background_steps)
    tags = list(feature_tags) + list(scenario['tags'])

    for step in scenario['steps']:
        steps.append(_pickle_step(step, path))

    pickle = {
        'tags': _pickle_tags(tags, path),
        'name': scenario['name'],
        'locations': [_pickle_location(scenario['location'], path)],
        'steps': steps,
        'type': 'pickle'
    }
    pickles.append(pickle)


def _compile_scenario_outline(feature_tags, background_steps, scenario_outline,
                              path, pickles):
    if len(scenario_outline['steps']) == 0:
      return

    for examples in (e for e in scenario_outline['examples'] if 'tableHeader' in e):
        variable_cells = examples['tableHeader']['cells']

        for values in examples['tableBody']:
            value_cells = values['cells']
            steps = list(background_steps)
            tags = list(feature_tags) \
                + list(scenario_outline['tags']) \
                + list(examples['tags'])

            for scenario_outline_step in scenario_outline['steps']:
                step_text = _interpolate(
                    scenario_outline_step['text'],
                    variable_cells,
                    value_cells)
                arguments = _create_pickle_arguments(
                    scenario_outline_step.get('argument'),
                    variable_cells,
                    value_cells,
                    path)
                _pickle_step = {
                    'text': step_text,
                    'arguments': arguments,
                    'locations': [
                        _pickle_location(values['location'], path),
                        _pickle_step_location(scenario_outline_step, path)
                    ]
                }
                steps.append(_pickle_step)

            pickle = {
                'name': _interpolate(
                    scenario_outline['name'],
                    variable_cells,
                    value_cells),
                'steps': steps,
                'tags': _pickle_tags(tags, path),
                'locations': [
                    _pickle_location(values['location'], path),
                    _pickle_location(scenario_outline['location'], path)
                ],
                'type': 'pickle'
            }
            pickles.append(pickle)


def _create_pickle_arguments(argument, variables, values, path):
    result = []

    if not argument:
        return result

    if argument['type'] is 'DataTable':
        table = {'rows': []}
        for row in argument['rows']:
            cells = [
                {
                    'location': _pickle_location(cell['location'], path),
                    'value': _interpolate(cell['value'], variables, values)
                } for cell in row['cells']
            ]
            table['rows'].append({'cells': cells})
        result.append(table)

    elif argument['type'] is 'DocString':
        docstring = {
            'location': _pickle_location(argument['location'], path),
            'content': _interpolate(argument['content'], variables, values)
        }
        result.append(docstring)

    else:
        raise Exception('Internal error')

    return result


def _interpolate(name, variable_cells, value_cells):
    for n, variable_cell in enumerate(variable_cells):
        value_cell = value_cells[n]
        name = re.sub(
            u'<{0[value]}>'.format(variable_cell),
            value_cell['value'],
            name
            )
    return name


def _pickle_steps(scenario_definition, path):
    return [_pickle_step(step, path)
        for step in scenario_definition['steps']]


def _pickle_step(step, path):
    return {
        'text': step['text'],
        'arguments': _create_pickle_arguments(
            step.get('argument'),
            [],
            [],
            path),
        'locations': [_pickle_step_location(step, path)]
    }


def _pickle_step_location(step, path):
    return {
        'path': path,
        'line': step['location']['line'],
        'column': step['location']['column'] + count_symbols(step.get('keyword', 0))
    }


def _pickle_location(location, path):
    return {
        'path': path,
        'line': location['line'],
        'column': location['column']
    }


def _pickle_tags(tags, path):
    return [_pickle_tag(tag, path) for tag in tags]


def _pickle_tag(tag, path):
    return {
        'name': tag['name'],
        'location': _pickle_location(tag['location'], path)
    }

File: dotnet/Gherkin/Pickles/Compiler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Gherkin.Ast;
// ReSharper disable PossibleMultipleEnumeration

namespace Gherkin.Pickles
{
    public class Compiler
    {
        public List<Pickle> Compile(GherkinDocument gherkinDocument, string path)
        {
            var pickles = new List<Pickle>();
            Feature feature = gherkinDocument.Feature;
            if (feature == null)
            {
                return pickles;
            }

            var featureTags = feature.Tags;
            var backgroundSteps = new PickleStep[0];

            foreach (var scenarioDefinition in feature.Children)
            {
                if (scenarioDefinition is Background)
                {
                    backgroundSteps = PickleSteps(scenarioDefinition, path);
                }
                else if (scenarioDefinition is Scenario)
                {
                    CompileScenario(pickles, backgroundSteps, (Scenario)scenarioDefinition, featureTags, path);
                }
                else {
                    CompileScenarioOutline(pickles, backgroundSteps, (ScenarioOutline)scenarioDefinition, featureTags, path);
                }
            }
            return pickles;
        }

        protected virtual void CompileScenario(List<Pickle> pickles, IEnumerable<PickleStep> backgroundSteps, Scenario scenario, IEnumerable<Tag> featureTags, string path)
        {
            if (!scenario.Steps.Any())
                return;

            var steps = new List<PickleStep>();
            steps.AddRange(backgroundSteps);

            var scenarioTags = new List<Tag>();
            scenarioTags.AddRange(featureTags);
            scenarioTags.AddRange(scenario.Tags);

            steps.AddRange(PickleSteps(scenario, path));

            Pickle pickle = CreatePickle(
                    scenario,
                    scenario.Name,
                    steps,
                    PickleTags(scenarioTags, path),
                    SingletonList(PickleLocation(scenario.Location, path))
            );
            pickles.Add(pickle);
        }

        protected virtual IEnumerable<T> SingletonList<T>(T item)
        {
            return new[] { item };
        }

        protected virtual void CompileScenarioOutline(List<Pickle> pickles, IEnumerable<PickleStep> backgroundSteps, ScenarioOutline scenarioOutline, IEnumerable<Tag> featureTags, string path)
        {
            if (!scenarioOutline.Steps.Any())
                return;

            foreach (var examples in scenarioOutline.Examples)
            {
                if (examples.TableHeader == null) continue;
                var variableCells = examples.TableHeader.Cells;
                foreach (var values in examples.TableBody)
                {
                    var valueCells = values.Cells;

                    var steps = new List<PickleStep>();
                    steps.AddRange(backgroundSteps);

                    var tags = new List<Tag>();
                    tags.AddRange(featureTags);
                    tags.AddRange(scenarioOutline.Tags);
                    tags.AddRange(examples.Tags);

                    foreach(var scenarioOutlineStep in scenarioOutline.Steps)
                    {
                        string stepText = Interpolate(scenarioOutlineStep.Text, variableCells, valueCells);

                        // TODO: Use an Array of location in DataTable/DocString as well.
                        // If the Gherkin AST classes supported
                        // a list of locations, we could just reuse the same classes

                        PickleStep pickleStep = CreatePickleStep(
                                scenarioOutlineStep,
                                stepText,
                                CreatePickleArguments(scenarioOutlineStep.Argument, variableCells, valueCells, path),
                                new[] {
                                        PickleLocation(values.Location, path),
                                        PickleStepLocation(scenarioOutlineStep, path)
                                }
                        );
                        steps.Add(pickleStep);
                    }

                    Pickle pickle = CreatePickle(
                            scenarioOutline,
                            Interpolate(scenarioOutline.Name, variableCells, valueCells),
                            steps,
                            PickleTags(tags, path),
                            new[] {
                                    PickleLocation(values.Location, path),
                                    PickleLocation(scenarioOutline.Location, path)
                            }
                    );

                    pickles.Add(pickle);
                }
            }
        }

        protected virtual Pickle CreatePickle(ScenarioDefinition scenarioDefinition, string name, IEnumerable<PickleStep> steps, IEnumerable<PickleTag> tags, IEnumerable<PickleLocation> locations)
        {
            return new Pickle(name, steps, tags, locations);
        }

        protected virtual PickleStep CreatePickleStep(Step step, string text, IEnumerable<Argument> arguments, IEnumerable<PickleLocation> locations)
        {
            return new PickleStep(text, arguments, locations);
        }

        protected virtual List<Argument> CreatePickleArguments(StepArgument argument, string path)
        {
            var noCells = Enumerable.Empty<TableCell>();
            return CreatePickleArguments(argument, noCells, noCells, path);
        }

        protected virtual List<Argument> CreatePickleArguments(StepArgument argument, IEnumerable<TableCell> variableCells, IEnumerable<TableCell> valueCells, string path)
        {
            var result = new List<Argument>();
            if (argument == null) return result;
            if (argument is DataTable) {
                DataTable t = (DataTable)argument;
                var rows = t.Rows;
                var newRows = new List<PickleRow>(rows.Count());
                foreach(var row in rows)
                {
                    var cells = row.Cells;
                    var newCells = new List<PickleCell>();
                    foreach(var cell in cells)
                    {
                        newCells.Add(
                                new PickleCell(
                                        PickleLocation(cell.Location, path),
                                        Interpolate(cell.Value, variableCells, valueCells)
                                )
                        );
                    }
                    newRows.Add(new PickleRow(newCells));
                }
                result.Add(new PickleTable(newRows));
            } else if (argument is DocString) {
                DocString ds = (DocString)argument;
                result.Add(
                        new PickleString(
                                PickleLocation(ds.Location, path),
                                Interpolate(ds.Content, variableCells, valueCells)
                        )
                );
            } else {
                throw new InvalidOperationException("Unexpected argument type: " + argument);
            }
            return result;
        }

        protected virtual PickleStep[] PickleSteps(ScenarioDefinition scenarioDefinition, string path)
        {
            var result = new List<PickleStep>();
            foreach(var step in scenarioDefinition.Steps)
            {
                result.Add(PickleStep(step, path));
            }
            return result.ToArray();
        }

        protected virtual PickleStep PickleStep(Step step, string path)
        {
            return CreatePickleStep(
                    step,
                    step.Text,
                    CreatePickleArguments(step.Argument, path),
                    SingletonList(PickleStepLocation(step, path))
            );
        }

        protected virtual string Interpolate(string name, IEnumerable<TableCell> variableCells, IEnumerable<TableCell> valueCells)
        {
            int col = 0;
            foreach (var variableCell in variableCells)
            {
                TableCell valueCell = valueCells.ElementAt(col++);
                string header = variableCell.Value;
                string value = valueCell.Value;
                name = name.Replace("<" + header + ">", value);
            }
            return name;
        }

        protected virtual PickleLocation PickleStepLocation(Step step, string path)
        {
            return new PickleLocation(
                    path,
                    step.Location.Line,
                    step.Location.Column + (step.Keyword != null ? step.Keyword.Length : 0)
            );
        }

        protected virtual PickleLocation PickleLocation(Location location, string path)
        {
            return new PickleLocation(path, location.Line, location.Column);
        }

        protected virtual List<PickleTag> PickleTags(List<Tag> tags, string path)
        {
            var result = new List<PickleTag>();
            foreach(var tag in tags)
            {
                result.Add(PickleTag(tag, path));
            }
            return result;
        }

        protected virtual PickleTag PickleTag(Tag tag, string path)
        {
            return new PickleTag(PickleLocation(tag.Location, path), tag.Name);
        }
    }
}
